### PR TITLE
Add cron to purge closed tickets

### PR DESCRIPTION
**Pull request by trasher**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #6207

### CODE REVIEW

File: inc/ticket.class.php [hunk 1]
```diff
@@ -6422,6 +6425,82 @@ static function cronCreateInquest($task) {
...
+            $criteria = [
+               'FROM'   => $ticket->getTable(),
+               'WHERE'  => [
+                  'entities_id'  => $entity['id'],
+                  'status'       => $ticket->getClosedStatusArray(),
+               ]
+            ];
+
+            if ($delay > 0) {
+               $calendars_id = Entity::getUsedConfig('calendars_id', $entity['id']);
```
Comment (flegastelois) [inline inc/ticket.class.php]: I think this is not necessary, we want a purge of all tickets whose **closedate** (DB value) is older than X days.

File: inc/ticket.class.php [hunk 2]
```diff
@@ -6204,6 +6204,9 @@ static function cronInfo($name) {
 
          case 'createinquest' :
             return ['description' => __('Generation of satisfaction surveys')];
+
+         case 'purgeticket':
+            return ['description' => __('Automatic tickets purge')];
```
Comment (flegastelois) [inline inc/ticket.class.php]:
```suggestion
            return ['description' => __('Automatic closed tickets purge and their attached documents')];
```

File: inc/ticket.class.php [hunk 3]
```diff
@@ -6422,6 +6425,82 @@ static function cronCreateInquest($task) {
...
+                  // no calendar, remove all days
+                  $criteria['WHERE']['closedate'] = new \QueryExpression("ADDDATE(`closedate`, INTERVAL ".$delay." DAY) < NOW()");
+               }
+            }
+
+            $iterator = $DB->request($criteria);
+            $nb = 0;
+
+            foreach ($iterator as $tick) {
+               $ticket->delete(
```
Comment (flegastelois) [inline inc/ticket.class.php]: > automatic action to purge closed tickets **and their attached documents**

It misses attached documents (DB item + storage file) purge, no?

Comment (trasher) [reply inc/ticket.class.php]: I think this is done along with ticket deletion, I'll check

Comment (trasher) [reply inc/ticket.class.php]: As discussed, I'm going to add another crontask in order to clean orphaned documents

File: inc/document.class.php [hunk 4]
```diff
@@ -1648,4 +1648,66 @@ public static function getImage($path, $context, $mwidth = null, $mheight = null
...
+
+      switch ($name) {
+         case 'cleanorphans' :
+            return ['description' => __('Clean orphaned documents')];
+      }
+      return [];
+   }
+
+   /**
+    * Cron for ticket's automatic purge
```
Comment (flegastelois) [inline inc/document.class.php]:
```suggestion
    * Cron for clean orphan documents (without Document_Item)
```

File: inc/ticket.class.php [hunk 5]
```diff
@@ -6204,6 +6204,9 @@ static function cronInfo($name) {
 
          case 'createinquest' :
             return ['description' => __('Generation of satisfaction surveys')];
+
+         case 'purgeticket':
+            return ['description' => __('Automatic closed tickets purge and their attached documents')];
```
Comment (flegastelois) [inline inc/ticket.class.php]:
```suggestion
            return ['description' => __('Automatic closed tickets purge')];
```

File: inc/document.class.php [hunk 6]
```diff
@@ -1648,4 +1648,66 @@ public static function getImage($path, $context, $mwidth = null, $mheight = null
       );
       return ($result ? $context_path : $path);
    }
+
+   /**
+    * Give cron information
+    *
+    * @param string $name task's name
+    *
+    * @return arrray of information
```
Comment (cedric-anne) [inline inc/document.class.php]:
```suggestion
    * @return array of information
```

File: inc/document.class.php [hunk 7]
```diff
@@ -1648,4 +1648,66 @@ public static function getImage($path, $context, $mwidth = null, $mheight = null
...
+   }
+
+   /**
+    * Cron for clean orphan documents (without Document_Item)
+    *
+    * @param Crontask $task Crontask object
+    *
+    * @return integer (0 : nothing done - 1 : done)
+   **/
+   static function cronCleanOrphans(Crontask $task) {
```
Comment (cedric-anne) [inline inc/document.class.php]: Problem with this is that it will delete all documents added directly from the `/front/document.form.php` form. As it is possible to create documents that are not related to any item, it is pretty sure that this feature is used by some users.

File: install/update_94_95.php [hunk 8]
```diff
@@ -452,6 +452,37 @@ function update94to95() {
...
+      );
+   }
+
+   CronTask::Register(
+      'Ticket',
+      'purgeticket',
+      7 * DAY_TIMESTAMP,
+      [
+         'mode'  => CronTask::MODE_EXTERNAL,
+         'state' => CronTask::STATE_DISABLE
```
Comment (cedric-anne) [inline install/update_94_95.php]: If it is disabled by default, admins that change the entity configuration may not understand why tickets are not purged. I think this cron should be active by default, even if it has nothing to do.

File: inc/document.class.php (hunk 7 above)
Comment (trasher) [reply inc/document.class.php]: well, if they want some documents that are not related to anything, they should just not enable this crontask

File: install/update_94_95.php (hunk 8 above)
Comment (trasher) [reply install/update_94_95.php]: I guess not, this may be destructive; admin have to enable such features explicitely.

Comment (orthagh) [reply install/update_94_95.php]: maybe a warning in the conf to alert the administrator the automatic is disabled (as the configuration and the action are separated) is this case, no ? Btw, i'm agree the action should be disabled by default

Comment (trasher) [reply install/update_94_95.php]: Message added
